### PR TITLE
chore: Migrating dashboard/components/menu from jsx to tsx

### PR DESCRIPTION
**Pull request by a5net**
### SUMMARY

Migrated files under `superset-frontend/dashboard/components/menu` from jsx to tsx.

#### Before and After for MarkdownModeDropdown and PopoverDropdown

Screen Shot 2021-02-26 at 20 03 21
Screen Shot 2021-02-26 at 19 27 28

#### Before and After for BackgroundStyleDropdown and PopoverDropdown

Screen Shot 2021-02-26 at 20 03 52
Screen Shot 2021-02-26 at 19 49 26

### ADDITIONAL INFORMATION

- [x] Has associated issue: #10004
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/dashboard/components/menu/WithPopoverMenu.tsx [hunk 1]
```diff
@@ -121,21 +136,17 @@ class WithPopoverMenu extends React.PureComponent {
...
-          <div className="popover-menu">
-            {menuItems.map((node, i) => (
-              <div className="menu-item" key={`menu-item-${i}`}>
-                {node}
-              </div>
-            ))}
-          </div>
-        )}
+        {editMode && isFocused && menuItems != undefined
+          && menuItems.length && menuItems.length > 0 && (
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/menu/WithPopoverMenu.tsx:140]: I think this could be simpler, something like:
```typescript
editMode && isFocused && (menuItems?.length ?? 0) > 0
```

File: superset-frontend/src/dashboard/components/menu/BackgroundStyleDropdown.tsx [hunk 2]
```diff
@@ -17,35 +17,36 @@
...
+import PopoverDropdown, { OptionProps } from './PopoverDropdown';
 
-const propTypes = {
-  id: PropTypes.string.isRequired,
-  value: PropTypes.string.isRequired,
-  onChange: PropTypes.func.isRequired,
+interface BackgroundStyleDropdownProps {
+  id: string,
+  value: string,
+  onChange: Function,
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/menu/BackgroundStyleDropdown.tsx:28]: can we define this function with arguments and return types too?

File: superset-frontend/src/dashboard/components/menu/HoverMenu.tsx [hunk 3]
```diff
@@ -16,23 +16,22 @@
...
+import React, { RefObject } from 'react';
 import cx from 'classnames';
 
-const propTypes = {
-  position: PropTypes.oneOf(['left', 'top']),
-  innerRef: PropTypes.func,
-  children: PropTypes.node,
+interface HoverMenuProps {
+  position?: 'left' | 'top',
+  innerRef?: RefObject<any>,
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/menu/HoverMenu.tsx:24]: we're trying to avoid using `any`, maybe:
- See if you can find the type here
- See if `unknown` works

And apply this principle to all the other cases of `any` too, thanks!

File: superset-frontend/src/dashboard/components/menu/MarkdownModeDropdown.tsx [hunk 4]
```diff
@@ -17,15 +17,14 @@
...
 import PopoverDropdown from './PopoverDropdown';
 
-const propTypes = {
-  id: PropTypes.string.isRequired,
-  value: PropTypes.string.isRequired,
-  onChange: PropTypes.func.isRequired,
+interface MarkdownModeDropdownProps {
+  id: string,
+  value: string,
+  onChange: Function,
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/menu/MarkdownModeDropdown.tsx:27]: same comment about function types

File: superset-frontend/src/dashboard/components/menu/PopoverDropdown.tsx [hunk 5]
```diff
@@ -75,20 +71,30 @@ const MenuItem = styled(Menu.Item)`
   }
 `;
 
-class PopoverDropdown extends React.PureComponent {
-  constructor(props) {
+interface HandleSelectProps {
+  key: any;
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/menu/PopoverDropdown.tsx:75]: this is probably a string or `string | int`?

File: superset-frontend/src/dashboard/components/menu/PopoverDropdown.tsx [hunk 6]
```diff
@@ -111,22 +116,19 @@ class PopoverDropdown extends React.PureComponent {
                   active: option.value === value,
                 })}
               >
-                {renderOption(option)}
+                {renderOption!(option)}
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/menu/PopoverDropdown.tsx:119]: any way to not need the bang here?

File: superset-frontend/src/dashboard/components/menu/PopoverDropdown.tsx [hunk 7]
```diff
@@ -98,8 +104,7 @@ class PopoverDropdown extends React.PureComponent {
     const selected = options.find(opt => opt.value === value);
     return (
       <Dropdown
-        id={id}
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/menu/PopoverDropdown.tsx:101]: was removing this intentional?

File: superset-frontend/src/dashboard/components/menu/WithPopoverMenu.tsx [hunk 8]
```diff
@@ -17,44 +17,54 @@
...
-  style: PropTypes.object,
+interface WithPopoverMenuProps {
+  children?: React.ReactNode,
+  disableClick?: Boolean,
+  menuItems?: React.ReactNode[],
+  onChangeFocus?: Function,
+  isFocused?: Boolean,
+  shouldFocus?: Function,
+  editMode: Boolean,
+  style?: Object,
```
Comment (etr2460) [inline superset-frontend/src/dashboard/components/menu/WithPopoverMenu.tsx:30]: Prefer `Record<string, any>`, or more specific types if possible

**Review by etr2460 — COMMENTED** [review]
Thanks for the contribution! A few comments about the types, and also I think you need to fix a conflict with the package-lock file

File: superset-frontend/src/dashboard/components/menu/PopoverDropdown.tsx (hunk 7 above)
Comment (ayanginet) [reply superset-frontend/src/dashboard/components/menu/PopoverDropdown.tsx:101]: Yes, because Dropdown does not have such props.

File: superset-frontend/src/dashboard/components/menu/PopoverDropdown.tsx (hunk 5 above)
Comment (ayanginet) [reply superset-frontend/src/dashboard/components/menu/PopoverDropdown.tsx:75]: I tried to look for when the `key` is used to determine the type. Likely, this is used by antd and the type is `React.Key`.

File: superset-frontend/src/dashboard/components/menu/WithPopoverMenu.tsx [hunk 9]
```diff
@@ -17,44 +17,58 @@
...
+
+  container: ShouldFocusContainer;
+
+  static defaultProps = {
+    children: null,
+    disableClick: false,
+    onChangeFocus: null,
+    menuItems: [],
+    isFocused: false,
+    shouldFocus: (event: any, container: ShouldFocusContainer) =>
```
Comment (ayanginet) [inline superset-frontend/src/dashboard/components/menu/WithPopoverMenu.tsx:55]: I tried to replace the type of the event from `any` to `React.FocusEvent<>` so I could have properties like target. 

However, if I do so, it raises another issue in handleClick(event). Here, both events have to be the same type. In handleClick(event) the function calls document.addEventListener('click', handleClick). addEventListener expects the following types: `(type: string, listener: EventListenerOrEventListenerObject, ...)`.

So basically, I ended up in a situation when I need to choose one from, but if I choose one the other breaks.

Comment (etr2460) [reply superset-frontend/src/dashboard/components/menu/WithPopoverMenu.tsx:53]: Hmm, yeah that seems tricky. Maybe add a summary of the issue as a comment, so that way a future engineer knows why `any` was used?

**Review by etr2460 — COMMENTED** [review]
Thanks for the iteration, this looks great now! Only one other item to address (adding a comment in the code), otherwise it lgtm

**Review by etr2460 — APPROVED** [review]

**Comment by etr2460** [comment]
Thanks for the contribution!
